0000-scheduled-circuit.md

```diff
@@ -0,0 +1,266 @@
+  - O(N): check validity of arguments
+  - O(N): update start times
+  - O(ListReplace(D)\*W): replace items on lists by qubit
+
+3) insert(ref: Position, targets: LengthedInstructions)*
+    - fail if no enough idle time before the reference instruction
+- Relatively-timed:
+  - O(W): check validity of arguments
+  - O(ListInsert(D)*W): insert items on lists by qubit
+- Absolutely-timed:
+  - O(W): check validity of arguments
+  - O(N): update start times
+  - O(ListInsert(D)*W): insert items on lists by qubit
+
+4) insert_at_time(time: Time, targets: LengthedInstructions)
+    - fail if no enough idle time between the time and start time of next instruction
+- Relatively-timed:
+  - O(D*W): overlap check
+  - O(ListReplace(D)*W): replace items on lists by qubit
+- Absolutely-timed:
+  - O(log(D)*W): overlap check
+  - O(N): update start times
+  - O(ListReplace(D)*W): replace items on lists by qubit
+
+In addition to these, we need kinds of “find()” family for users to specify references in insert/replace etc.
+
+\*) I assume the interface that references are given as Postions.
+There is a clear drawback in referring by positions:
+replacement of cx gate, which would be specified by two Positions and may confuse users.
+(*To achieve single position for cx gate, it would be better to use DAG data structure like DAGCircuit…*).
+If references should be given as Instructions, we need to accept overhead to find positions.
+Such overhead would be O(D\*W) for Relatively-timed and O(log(D)*W) for Absolutely-timed unless we add another data structure for the purpose.
+
+
+### Dictionary vs DAG data structure
+How to store *ordering* among instructions is much important point for selecting data structure.
+This is a problem independent of Relatively-timed or Absolutely-timed, which addresses the problem if we should keep absolute-timing in data or not.
+
+We can think of two data structure for storing the orders among instructions:
+- Dictionary (Dict[Qubit, List[LengthedInstruction]] discussed above)
+- DAG (as like used in DAGCircuit)
+
+Note: *For any node of DAGCircuit, its in-(or out-)edges must have different qubit label.
+It enables DAGCircuit to store the ordering of instruction by qubits.*
+As for general DependencyDAG, which represents dependencies among instructions,
+its node can have multiple in-(or out-)-edges with the same qubit label.
+So DependencyDAG is not appropriate to store the ordering of instruction by qubits.
+
+In Dictionary, any multi-qubit instruction is stored at multiple positions on multiple lists.
+So referring a multi-qubit instruction with one position is not possible.
+In DAG, it is possible because any node can be specified by "instruction id (an integer)".
+To get instruction by its id, we may need to store a map from id to instruction (as list or dict).
+
+
+
+## Questions
+- any other use case? noise simulation with custom gate (with custom duration and noise)?
+
+- need to support "insert_at_time()"?
```

Comment:
I believe so. Handling overlaps with floating point time will be more nuanced and error prone.